File: solutions/python/03/binary_search_answer/main.py
import sys


def main() -> None:
    n = int(sys.stdin.readline())

    def check(x: int) -> bool:
        return x * x <= n

    hi = 1
    while check(hi):
        hi *= 2
    lo, hi_ans = 0, hi
    while lo < hi_ans:
        mid = (lo + hi_ans + 1) // 2
        if check(mid):
            lo = mid
        else:
            hi_ans = mid - 1
    print(lo)


if __name__ == "__main__":
    main()

File: solutions/go/03/binary_search_answer/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(in, &n)
	check := func(x int) bool { return x*x <= n }
	hi := 1
	for check(hi) {
		hi *= 2
	}
	lo, hiAns := 0, hi
	for lo < hiAns {
		mid := (lo + hiAns + 1) / 2
		if check(mid) {
			lo = mid
		} else {
			hiAns = mid - 1
		}
	}
	fmt.Println(lo)
}
